### PR TITLE
Use asserts instead of exceptions for executor not started

### PR DESCRIPTION
**Pull request by dstandish**
It's a little cleaner, and I'm pretty sure the actual practical import here is just getting mypy to cooperate, so I think the intention is clearer this way too.

## question

do we really need the "not started" message? it's ugly...

### CODE REVIEW

**Comment by potiuk** [comment]
While I heartily sympathise with you, currently we agreed to not use asserts in the main code. We should only use then in tests.

Initial Discussion here: https://lists.apache.org/thread/h22sodtylksn04tzm1f2n3q8h994j38p
Voting thread here: https://lists.apache.org/thread/tl9f344vwjmzh8pqfgbjkx8rqyy7dp9n

I think by disabling pylint (another vote) we lost check for it, it used to be flagged in pre-commits (and if we are going to keep the rule, we likely should add it). 

As you will see, the proponent of using asserts (myself including) were out-voted, and we had no other choice by follow it (I even implemented pylint check for that). 

Main reason why we (as community decided to forbid it, was potential confusion about the flows 

But - maybe it's time to re-open the discussion ? I am still of an opinion that asserts in cases like you mention are rather useful. As we say - "disagree, but engage" - but if we vote again a decision is different, I am super happy to engage on the side of asserts.

**Comment by potiuk** [comment]
> There is an exception for type checking though, so if these are truly only for mypy...
> 
> https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#dont-use-asserts-outside-tests

Ah yeah.  Forgot about it ... And I have not linked the CONTRIBUTING section 🤦 . Thanks @jedcunningham for that one. Refreshed my memory about it.

**Comment by dstandish** [comment]
> > There is an exception for type checking though, so if these are truly only for mypy...
> > https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#dont-use-asserts-outside-tests
> 
> Ah yeah. Forgot about it ... And I have not linked the CONTRIBUTING section 🤦 . Thanks @jedcunningham for that one. Refreshed my memory about it.

Yeah so.... This is for type checking afaict

Is it really necessary to add the if?

**Comment by dstandish** [comment]
OK -- I updated to use the `if TYPE_CHECKING` style

File: airflow/executors/base_executor.py [hunk 1]
```diff
@@ -33,8 +33,6 @@
 
 PARALLELISM: int = conf.getint("core", "PARALLELISM")
 
-NOT_STARTED_MESSAGE = "The executor should be started first!"
-
```
Comment (dstandish) [inline airflow/executors/base_executor.py:37]: maybe we need to maintain this for backcompat but that's basically xkcd1172

**Comment by potiuk** [comment]
> OK -- I updated to use the `if TYPE_CHECKING` style

I think this is misunderstanding. This "if TYPE_CHECKING" is not something that you should add. There are specific places where we want to use assert for `if TYPE_CHECKING` when we want to get rid of `mypy` warning that None has no property etc. The role of "assert" in tnis case is to "calm down" MyPy so that it does not complain, in places where we **know** None cannot hapen for example.

But in the case above, we actually want raise those messages if something unexpected happens. The way  it is done now - those messages will never be printed because if TYPE_CHECKING is only true when MyPy parses the code.

I think - with the current state of our approach - you simply can't replace those if-s  wiht asserts.

**Comment by dstandish** [comment]
> we want to use assert for if TYPE_CHECKING when we want to get rid of mypy warning that None has no property etc

Yes, and by my reading, it appears that's exactly what these asserts are there for.  Essentially what's required is that executor.start is called before the executor is used.  And, it looks like that's in fact what happens.  

It looks like you added these initially [here](https://github.com/apache/airflow/commit/a36cfe049a2c5948b24fde7a878fe19cabede5f7).

Presumably, prior to that change, mypy never complained because the attrs were not defined and annotated in the init method -- they were created in `start`.  But in your change you initialized them as None and annotated them, e.g. [here](https://github.com/apache/airflow/blob/a36cfe049a2c5948b24fde7a878fe19cabede5f7/airflow/executors/dask_executor.py#L45).  I presume that this is why the asserts had to be added (e.g. [here](https://github.com/apache/airflow/blob/a36cfe049a2c5948b24fde7a878fe19cabede5f7/airflow/executors/dask_executor.py#L66)) -- to get mypy to not complain.

That PR I assume predated any discussions about asserts vs raising.  Then when it was decided to go with raising, except in the case of mypy appeasement, these asserts were changed.  But, it does seem that these are mypy appeasement and could therefore have been left alone, or put behind a type checking guard, as I've proposed.  

In any case, if you see it another way, that's fine and you're welcome to close this.

**Comment by potiuk** [comment]
Ah ok. Now I See. Yeah. I think I indeed added it for `None` mypy case when we had not yet agreed asserts case. Makes perfect sense. I thought it was more "flow" protection, but yeah. This way is better.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
BTW. the if TYPE/assert is indeed much better in this case as it shows intention. The previous implementation (of old mine) tricked the future mine into thinking that it was somewhat possible to get into this flow. 

Funny thing is that In a way "just" assert wouldn't be as good as "if'd" one for similar reasons - it would not have clear indication that this assert is there only beacuse of MyPy. But now it's crystal clear.

**Pull request merged**